Instagram/ViewController/HomeFeedViewController.m

```diff
@@ -40,22 +58,110 @@ - (IBAction)didTapLogout:(id)sender {
 - (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
     PostCell *cell = [tableView dequeueReusableCellWithIdentifier:@"PostCell" forIndexPath:indexPath];
     Post *post = self.arrayOfPosts[indexPath.row];
-    NSURL *url = [NSURL URLWithString:post.picture];
-    NSString *newImage = [[NSString alloc] initWithContentsOfURL:url
-                                           encoding:NSUTF8StringEncoding error:nil];
-    cell.postImage.image = [UIImage imageNamed:newImage];
+    [post.image getDataInBackgroundWithBlock:^(NSData *data, NSError *error) {
+        cell.postImage.image = [UIImage imageNamed:@"placeholder.png"]; // placeholder image
+        cell.postImage.image = [UIImage imageWithData:data];
+        cell.userInteractionEnabled = YES;
+    }];
     cell.captionLabel.text = post.caption;
     cell.post = post;
+    cell.selectionStyle = UITableViewCellSelectionStyleNone;
     return cell;
 }
 
 - (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
     return self.arrayOfPosts.count;
 }
 
-- (void)didPost:(nonnull Post *)post {
-    [self.arrayOfPosts insertObject:post atIndex:0];
-    [self.tableView reloadData];
+- (void)didPost:(Post *)post {
+    // construct query
+    PFQuery *query = [PFQuery queryWithClassName:@"Post"];
+    [query orderByDescending:@"createdAt"];
+    [query includeKey:@"author"];
+    query.limit = 20;
+
+    // fetch data asynchronously
+    [query findObjectsInBackgroundWithBlock:^(NSArray *posts, NSError *error) {
+        if (posts != nil) {
+            self.arrayOfPosts = (NSMutableArray *)posts;
+            [self.tableView reloadData];
+        }
+    }];
+}
+
+// Makes a network request to get updated data
+// Updates the tableView with the new data
+// Hides the RefreshControl
+- (void)beginRefresh:(UIRefreshControl *)refreshControl {
+    // load posts
+    PFQuery *query = [PFQuery queryWithClassName:@"Post"];
+    [query orderByDescending:@"createdAt"];
+    [query includeKey:@"author"];
+    query.limit = 20;
+
+    // fetch data asynchronously
+    [query findObjectsInBackgroundWithBlock:^(NSArray *posts, NSError *error) {
+        if (posts != nil) {
+            self.arrayOfPosts = (NSMutableArray *)posts;
+            [self.tableView reloadData];
+        }
+        // Tell the refreshControl to stop spinning
+        [refreshControl endRefreshing];
+    }];
+}
+
+- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
+    if(!self.isMoreDataLoading){
```

Comment:
This is fine since we are only making one api call. However, I personally like to do this in a dispatch queue. Please take a look at them when you have some free time! https://developer.apple.com/documentation/dispatch/dispatchqueue